### PR TITLE
[mle] add API for changing number of child router links

### LINKED ISSUE #8325: Low multicast delivery rates in big matter networks
**Describe the bug**

We have been testing multicast reliability in a network of approx. 50 Nordic DK with the matter lighting-app and matter's group functionality.

In this network we see an considerably higher loss rate of the multicast packets in REEDs compared to routers.

**To Reproduce**

1. Git commit id
NCS 2.0.2
commit aacd9eb0ba0d45483613370d8b7924fe3fa71606 (HEAD, manifest-rev)
Merge: e299df7fc f28a3e5ec
Author: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Date:   Mon Jul 25 15:56:27 2022 +0200

2. IEEE 802.15.4 hardware platform
nRF52840 DK

3. Build steps
Binary build as distributed by NCS2.0.2.

4. Network topology
Dense topology: All devices on a table. 

**Expected behavior** 
Similar loss rates for REEDs and Routers.

**Console/log output**
N/A

**Additional context**
On request of @LuDuda we have changed `OPENTHREAD_CONFIG_MLE_CHILD_ROUTER_LINKS` to 12 and built OpenThread from sources, which resulted in a considerably better loss rate.

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
Multicast has been proven to work unreliably with a low value of OPENTHREAD_CONFIG_MLE_CHILD_ROUTER_LINKS for big dense networks. Since there is no silver bullet value that suits all installations, make it possibly to configure its value.

Fixes #8325.

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

### CODE REVIEW

File: src/core/thread/mle_router.hpp [hunk 1]
```diff
@@ -325,6 +325,22 @@ class MleRouter : public Mle
...
+     */
+    uint8_t GetChildRouterLinks(void) const { return mChildRouterLinks; }
+
+    /**
+     * This method sets the MLE_CHILD_ROUTER_LINKS value.
+     *
+     * @param[in]  aChildRouterLinks  The MLE_CHILD_ROUTER_LINKS value.
+     *
+     */
+    void SetChildRouterLinks(uint8_t aChildRouterLinks) { mChildRouterLinks = aChildRouterLinks; }
```
Comment (kghost) [inline src/core/thread/mle_router.hpp:342]: Do we need to check whether current active link count is less than `aChildRouterLinks` here ?

File: src/core/thread/mle_types.hpp [hunk 2]
```diff
@@ -161,6 +161,8 @@ constexpr uint8_t kRouterSelectionJitter      = 120; ///< (in sec)
 constexpr uint8_t kRouterDowngradeThreshold = 23;
 constexpr uint8_t kRouterUpgradeThreshold   = 16;
 
+constexpr uint8_t kChildRouterLinks = OPENTHREAD_CONFIG_MLE_CHILD_ROUTER_LINKS;
```
Comment (kghost) [inline src/core/thread/mle_types.hpp:164]: I'm not sure this variable is useful, why not use `OPENTHREAD_CONFIG_MLE_CHILD_ROUTER_LINKS` directly.

**Review by abtink — COMMENTED** [review]
Thanks @markus-becker-tridonic-com. Looks great overall. 👍 

One idea related to https://github.com/openthread/openthread/pull/8410/files#r1028781734 to limit the API use to when MLE is disabled

File: src/core/thread/mle_types.hpp (hunk 2 above)
Comment (kghost) [reply src/core/thread/mle_types.hpp:164]: Thanks for the explaination.

**Comment by markus-becker-tridonic-com** [comment]
Adapted to upstream changes.

File: src/core/thread/mle_router.hpp (hunk 1 above)
Comment (abtink) [reply src/core/thread/mle_router.hpp]: Yes. Adding `    VerifyOrExit(instance.Get<Mle::MleRouter>().IsDisabled(), error = kErrorInvalidState);` to only allow param change when device is not running/attached.

File: src/core/thread/mle_router.hpp [hunk 3]
```diff
@@ -565,7 +581,6 @@ class MleRouter : public Mle
 
     static constexpr uint8_t kLinkRequestMinMargin    = OPENTHREAD_CONFIG_MLE_LINK_REQUEST_MARGIN_MIN;
     static constexpr uint8_t kPartitionMergeMinMargin = OPENTHREAD_CONFIG_MLE_PARTITION_MERGE_MARGIN_MIN;
-    static constexpr uint8_t kChildRouterLinks        = OPENTHREAD_CONFIG_MLE_CHILD_ROUTER_LINKS;
```
Comment (abtink) [inline src/core/thread/mle_router.hpp:568]: Suggest keeping this constant so we can use it when we initialize `mChildRouterLinks` in the constructor.

File: include/openthread/thread_ftd.h [hunk 4]
```diff
@@ -355,6 +355,29 @@ uint8_t otThreadGetRouterUpgradeThreshold(otInstance *aInstance);
...
+/**
+ * Set the MLE_CHILD_ROUTER_LINKS parameter used in the REED role.
+ *
+ * @param[in]  aInstance         A pointer to an OpenThread instance.
+ * @param[in]  aChildRouterLinks The MLE_CHILD_ROUTER_LINKS value.
+ *
+ * @sa otThreadGetChildRouterLinks
+ *
+ */
+void otThreadSetChildRouterLinks(otInstance *aInstance, uint8_t aChildRouterLinks);
```
Comment (abtink) [inline include/openthread/thread_ftd.h:379]: Suggest returning `otError` from this API:
```
 * @retval OT_ERROR_NONE            Successfully set the value.
 * @retval OT_ERROR_INVALID_STATE   Thread protocols are enabled.
 *

File: src/core/thread/mle_router.hpp (hunk 3 above)
Comment (markus-becker-tridonic-com) [reply src/core/thread/mle_router.hpp]: done.

File: include/openthread/thread_ftd.h (hunk 4 above)
Comment (markus-becker-tridonic-com) [reply include/openthread/thread_ftd.h]: done.

File: src/core/thread/mle_router.hpp (hunk 1 above)
Comment (markus-becker-tridonic-com) [reply src/core/thread/mle_router.hpp]: done.

File: include/openthread/thread_ftd.h [hunk 5]
```diff
@@ -355,6 +355,32 @@ uint8_t otThreadGetRouterUpgradeThreshold(otInstance *aInstance);
  */
 void otThreadSetRouterUpgradeThreshold(otInstance *aInstance, uint8_t aThreshold);
 
+/**
+ * Get the MLE_CHILD_ROUTER_LINKS parameter used in the REED role.
+ *
+ * @param[in]  aInstance A pointer to an OpenThread instance.
```
Comment (abtink) [inline include/openthread/thread_ftd.h:361]:
```suggestion
 * Get the MLE_CHILD_ROUTER_LINKS parameter used in the REED role.
 *
 * This parameter specifies the max number of neighboring routers with which the device (as an FED) 
 *  will try to establish link. *
 
 * @param[in]  aInstance A pointer to an OpenThread instance.
```

File: src/cli/cli.cpp [hunk 6]
```diff
@@ -6144,6 +6144,12 @@ template <> otError Interpreter::Process<Cmd("routerupgradethreshold")>(Arg aArg
 {
     return ProcessGetSet(aArgs, otThreadGetRouterUpgradeThreshold, otThreadSetRouterUpgradeThreshold);
 }
+
+template <> otError Interpreter::Process<Cmd("childrouterlinks")>(Arg aArgs[])
+{
+    return ProcessGetSet(aArgs, otThreadGetChildRouterLinks, otThreadSetChildRouterLinks);
+}
```
Comment (abtink) [inline src/cli/cli.cpp:6151]: Need to also add the `CmdEntry("childrouterlinks")` in `Interpreter::ProcessCommand()`.

**Review by abtink — COMMENTED** [review]
Thanks @markus-becker-tridonic-com. Looks good overall. 
Couple of smaller suggestions below:

File: src/core/thread/mle_router.hpp [hunk 7]
```diff
@@ -306,6 +306,30 @@ class MleRouter : public Mle
...
+     */
+    otError SetChildRouterLinks(uint8_t aChildRouterLinks)
+    {
+        otError error = OT_ERROR_NONE;
+        VerifyOrExit(IsDisabled(), error = kErrorInvalidState);
+
+        mChildRouterLinks = aChildRouterLinks;
+    exit:
+        return error;
+    }
```
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply src/core/thread/mle_router.hpp]: done

File: include/openthread/thread_ftd.h [hunk 8]
```diff
@@ -355,6 +355,35 @@ uint8_t otThreadGetRouterUpgradeThreshold(otInstance *aInstance);
  */
 void otThreadSetRouterUpgradeThreshold(otInstance *aInstance, uint8_t aThreshold);
 
+/**
+ * Get the MLE_CHILD_ROUTER_LINKS parameter used in the REED role.
+ *
+ * This parameter specifies the max number of neighboring routers with which the device (as an FED) 
+ *  will try to establish link. *
+
```
Comment (abtink) [inline include/openthread/thread_ftd.h:363]:
```suggestion
 * This parameter specifies the max number of neighboring routers with which the device (as an FED) 
 *  will try to establish link. 
 *
```

File: src/core/thread/mle_router.cpp [hunk 9]
```diff
@@ -3483,6 +3484,17 @@ void MleRouter::HandleAddressSolicitResponse(Coap::Message          *aMessage,
     InformPreviousChannel();
 }
 
+Error MleRouter::SetChildRouterLinks(uint8_t aChildRouterLinks)
+{
+    Error error = kErrorNone;
+    VerifyOrExit(IsDisabled(), error = kErrorInvalidState);
+
+    mChildRouterLinks = aChildRouterLinks;
```
Comment (abtink) [inline src/core/thread/mle_router.cpp:3492]:
```suggestion
    Error error = kErrorNone;
    
    VerifyOrExit(IsDisabled(), error = kErrorInvalidState);
    mChildRouterLinks = aChildRouterLinks;
```

File: src/core/thread/mle_router.hpp [hunk 10]
```diff
@@ -306,6 +306,22 @@ class MleRouter : public Mle
...
+     *
+     */
+    uint8_t GetChildRouterLinks(void) const { return mChildRouterLinks; }
+
+    /**
+     * This method sets the MLE_CHILD_ROUTER_LINKS value.
+     *
+     * @param[in]  aChildRouterLinks  The MLE_CHILD_ROUTER_LINKS value.
+     *
+     */
```
Comment (abtink) [inline src/core/thread/mle_router.hpp:322]:
```suggestion
     * @param[in]  aChildRouterLinks  The MLE_CHILD_ROUTER_LINKS value.
     *
     */
```
```suggestion
     * @param[in]  aChildRouterLinks  The MLE_CHILD_ROUTER_LINKS value.
     *
     * @retval kErrorNone          Successfully set the value.
     * @retval kErrorInvalidState  Thread protocols are enabled.
     */
```

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 👍 
Couple of small style suggestions below.

**Comment by markus-becker-tridonic-com** [comment]
@jwhui anything else missing on this PR?

**Review by jwhui — APPROVED** [review]

**Pull request merged**